### PR TITLE
Fix DateOnly and TimeOnly Formatting using interpolated strings

### LINKED ISSUE #64292: Invalid format specifier on DateOnly in string interpolation leads to infinite loop instead of FormatException
### Description

If you attempt to use an invalid format specifier on a `DateOnly` within string interpolation, instead of getting a `FormatException` like you would if you use an invalid format specifier on, for example, an int, instead it seems that an infinite loop happens along with massive memory usage and in some cases an `OutOfMemoryException`.

### Reproduction Steps

Easiest way to see the problem would probably be in LINQPad but could also be done in any place that can run .NET 6. The following line is all that is needed to see the problem occur:

`$"{new DateOnly(2022, 1, 25):u}"`

But the `ToString(string)` method of `DateOnly` does correctly throw a `FormatException`.

### Expected behavior

I expected to get a `FormatException` like I would if I tried to do `$"{1:a}"` for instance.

### Actual behavior

Actual behavior is that the code never finishes and memory usage balloons out of control (on my system, it would hover around 15GB of memory in use but going up and down as well).

### Regression?

Since `DateOnly` debuted with .NET 6, it could not be a regression.

### Known Workarounds

I know of no workarounds other than just not using an invalid format specifier in the first place.

### Configuration

.NET 6.0.1
Windows 10 64-bit 21H2 (build 19044.1469)
x64

I do not know if my issue is specific to this configuration, but from stepping through the code in a debugger, it does not seem like this configuration is the issue.

### Other information

I don't know of a fix but the problem seems to be partially due to the `TryFormat` method of `DateOnly` along with the `AppendFormatted` method of `DefaultInterpolatedStringHandler`. As far as I can tell, because `TryFormat` returns `false` and sets `charsWritten` to 0, .NET keeps trying to grow the buffer indefinitely.

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/64292

### CODE REVIEW

**Review by stephentoub — APPROVED** [review]
Thanks!  It's interesting this doesn't cause any tests to fail.  Are we missing direct tests of TryFormat?

**Comment by tarekgh** [comment]
>  It's interesting this doesn't cause any tests to fail. Are we missing direct tests of TryFormat?

We were testing the positive cases only. I added the interpolated string tests which should catch the negative cases. If you think we should add specific negative tests for TryFormat, I can do that.

**Comment by stephentoub** [comment]
Thanks. Yes, I think we should have those tests for TryFormat. It'd be fine to do those instead of the interpolation ones.

File: src/libraries/System.Runtime/tests/System/TimeOnlyTests.cs [hunk 1]
```diff
@@ -483,12 +483,15 @@ public static void TryFormatTest()
...
-        public static void InvalidFormattingWithInterpolatedStringTest()
-        {
-            TimeOnly timeOnly = TimeOnly.FromDateTime(DateTime.Now);
+            Assert.Throws<FormatException>(() => {
+                    Span<char> buff = stackalloc char[100];
+                    timeOnly.TryFormat(buff, out charsWritten, "u");
+                });
+            Assert.Throws<FormatException>(() => {
+                    Span<char> buff = stackalloc char[100];
+                    timeOnly.TryFormat(buff, out charsWritten, "dd-yyyy");
```
Comment (stephentoub) [inline src/libraries/System.Runtime/tests/System/TimeOnlyTests.cs:493]: Am I understanding correctly that there are some format specifiers that have special meaning, others that aren't recognized and are just treated as custom specifiers, and others that are considered erroneous and are failures?  Do we have sufficient test coverage for each of those categories?

Comment (tarekgh) [reply src/libraries/System.Runtime/tests/System/TimeOnlyTests.cs:493]: We have enough tests for formatting in general through testing with `ToString()` which cover all cases you mentioned. If you are talking specifically about TryFormat, we test with the `o` and `r` formats only. I was focusing here to add tests for the case I am fixing (as I am thinking of porting this fix to 6.0). but I can try adding such tests either now or in another PR.

Comment (stephentoub) [reply src/libraries/System.Runtime/tests/System/TimeOnlyTests.cs:493]: > We have enough tests for formatting in general through testing with ToString() which cover all cases you mentioned

Ok, thanks.

**Review by stephentoub — APPROVED** [review]

**Comment by tarekgh** [comment]
The failures in the CI is tracked by the issue https://github.com/dotnet/runtime/issues/64389.

**Pull request merged**